plasmapy/simulation/abstractions.py

```diff
@@ -0,0 +1,96 @@
+from abc import ABC, abstractmethod
+from typing import NoReturn, Union
+from numbers import Real, Integral
+from astropy import units as u
+
+
+class AbstractSimulation(ABC):
+    """
+    A prototype abstract interface for numerical simulations.
+
+    Notes
+    -----
+    This interface has unstable eigenfunctions, and is therefore subject
+    to change without notice.
+    """
+
+    @abstractmethod
+    def check_inputs(self) -> NoReturn:  # rename to validate_inputs?
+        """
+        Check the inputted simulation parameters.  Raise an exception or
+        issue a warning if any potential or actual problems are found.
+        """
+
+        pass
+
+    @abstractmethod
+    def summarize(self) -> str:
+        """
+        Return a string containing a summary of the simulation parameters
+        and status.
+        """
+
+        pass
+
+    @abstractmethod
+    def initialize(self) -> NoReturn:
+        """Prepare the simulation to be run."""
+
+        pass
+
+    @abstractmethod
+    def simulate(self) -> NoReturn:  # Return a class for the simulation output?
+        """Perform the actual simulation."""
+
+        pass
+
+    @abstractmethod
+    def finalize(self) -> NoReturn:
+        """Perform the steps to close the simulation and output data."""
+
+        pass
+
+
+class AbstractTimeDependentSimulation(AbstractSimulation):
+    """
+    A prototype abstract interface for time-dependent numerical simulations.
+
+    Notes
+    -----
+    This interface has unstable eigenfunctions, and is therefore subject
+    to change without notice.
+    """
+
+    @property
+    @abstractmethod
+    def start_time(self) -> Union[Real, u.Quantity]:
+        """The start time of the simulation."""
+
+        pass
+
+    @property
+    @abstractmethod
+    def end_time(self) -> Union[Real, u.Quantity]:
+        """The end time of the simulation."""
+
+        pass
+
+    @property
+    @abstractmethod
+    def time_step(self) -> Union[Real, u.Quantity]:
+        """Return the time step of the simulation."""
+
+        pass
+
+    @abstractmethod
+    def time_advance(self) -> NoReturn:
+        """Perform one step of the time advance."""
+
+        pass
```

Comment:
Putting these in as `abstractmethods` means they're required, and any simulation  we end up implementing must have this sort of logic:

```python
while time < self.end_time():
	self.time_step()
	self.time_advance()
```

For a simulation of, say, 1e10 steps, that's a lot of Python function calls and a lot of time we'd be spending just calling these two functions, whatever they are. Then we cannot wrap this in Numba:

```python
@numba.njit
def run(...):
    while time < end_time: 
        time_step(...)
        time_advance(...)
```

which, for some simulations, could be a massive time saver.

I'm aware that I changed the interface pretty seriously; this may be something we can avoid with [JITclass](https://numba.pydata.org/numba-doc/dev/user/jitclass.html), but I still haven't started figuring that out except for "it's not as good as it could be right now". Though this part seems new and promising:

> All methods of a jitclass are compiled into nopython functions. The data of a jitclass instance is allocated on the heap as a C-compatible structure so that any compiled functions can have direct access to the underlying data, bypassing the interpreter.

So maybe it's viable, doing it that way! I don't know just yet. Further research necessary.

Reply:
Yeah, those are good points.  There is also the use case that someone may wish to have the `simulate` or `run` method call a Fortran or C code and essentially do nothing in Python.  

I'm wondering if we would want to have a hierarchy of abstract base classes of varying specificity.  The most generic one could have the `initialize`, `simulate`, and `finalize` abstract methods and nothing else.  A more specific one could have `time_step`, `time_advance`, and so on.  Having them be functions in order to facilitate JIT compiling is an interesting thought, which is reminding me of Julia's approach.  I'll have to look into `numba.jitclass` too.  

In any case, everything I put down so far is in the brainstorm phase, as we'll need to have a few discussions to figure out what should go in there.  I'm also wondering if some of the things we want to include should be in a metadata schema rather than an abstract base class.  Hm...

Reply:
I did look into jitclass a bit and I think it's going to work with this kind of approach. I'll upload the notebook later and I can show you at the telecon

What we could do, though, is to limit the ABC part to a more user facing part of the interface. It's not necessary to allow single time steps to be run externally, though it'd be nice for debugging, I guess; but initial conditions, running the loop, data save and load and possibly restart from saved data (probably some more, too) would be pretty consistently necessary for simulations.

Reply:
I did look into jitclass a bit and I think it's going to work with this kind of approach. I'll upload the notebook later and I can show you at the telecon

What we could do, though, is to limit the ABC part to a more user facing part of the interface. It's not necessary to allow single time steps to be run externally, though it'd be nice for debugging, I guess; but initial conditions, running the loop, data save and load and possibly restart from saved data (probably some more, too) would be pretty consistently necessary for simulations.

Reply:
I've also been wondering about how the different ABCs will fit together.  I've been thinking for a while about how we need one each for initial conditions, boundary conditions, and the computational domain.  We also need one to allow us to work with simulation results.  

There are probably also going to be some components that users may not end up needing...like a data save and load may not be necessary for a short simulation that only gets interpreted interactively.

So many things to think about!  :scream_cat: